enderio-base/src/main/java/com/enderio/base/common/capability/darksteel/DarkSteelUpgradeable.java

```diff
@@ -0,0 +1,173 @@
+        Optional<IDarkSteelUpgradable> cap = is.getCapability(EIOCapabilities.DARK_STEEL_UPGRADABLE).resolve();
+        if(cap.isEmpty()) {
+            return Optional.empty();
+        }
+        return cap.get().getUpgradeAs(upgrade, as);
+    }
+
+    public static Collection<IDarkSteelUpgrade> getUpgradesThatCanBeAppliedAtTheMoment(ItemStack is) {
+        return is.getCapability(EIOCapabilities.DARK_STEEL_UPGRADABLE).map(IDarkSteelUpgradable::getUpgradesThatCanBeAppliedAtTheMoment).orElse(Collections.emptyList());
+    }
+
+    public static Collection<IDarkSteelUpgrade> getAllPossibleUpgrades(ItemStack is) {
+        return is.getCapability(EIOCapabilities.DARK_STEEL_UPGRADABLE).map(IDarkSteelUpgradable::getAllPossibleUpgrades).orElse(Collections.emptyList());
+    }
+
+    //----------------------- Class
+
+    private final Map<String, IDarkSteelUpgrade> upgrades = new HashMap<>();
+
+    /** The type of item that is upgradable, used to determine valid upgrades.*/
+    private ResourceLocation onItem;
+
+    public DarkSteelUpgradeable(){
+        this(EnderIO.loc("empty"));
+    }
+
+    public DarkSteelUpgradeable(ResourceLocation onItem) {
+        this.onItem = onItem;
+    }
+
+    @Override
+    public void addUpgrade(IDarkSteelUpgrade upgrade) {
+        removeUpgradeInSlot(upgrade.getSlot());
+        upgrades.put(upgrade.getSerializedName(), upgrade);
+    }
+
+    @Override
+    public void removeUpgrade(String name) {
+        upgrades.remove(name);
+    }
+
+    private void removeUpgradeInSlot(String slot) {
+        Optional<String> toRemove = Optional.empty();
+        for(var entry : upgrades.entrySet()) {
+            if(entry.getValue().getSlot().equals(slot)) {
+                toRemove = Optional.of(entry.getKey());
+                break;
+            }
+        }
+        toRemove.ifPresent(upgrades::remove);
+    }
+
+    @Override
+    public boolean canApplyUpgrade(IDarkSteelUpgrade upgrade) {
+        if(upgrades.isEmpty()) {
+            return EmpoweredUpgrade.NAME.equals(upgrade.getSerializedName()) && upgrade.isBaseTier();
+        }
+
+        Optional<IDarkSteelUpgrade> existing = getUpgrade(upgrade.getSerializedName());
+        if(existing.isPresent()) {
+            return existing.get().isValidUpgrade(upgrade);
+        }
+        if(!upgrade.isBaseTier()) {
+            return false;
+        }
+        return DarkSteelUpgradeRegistry.instance().getUpgradesForItem(onItem).contains(upgrade.getSerializedName());
+    }
+
+    @Override
+    public Optional<IDarkSteelUpgrade> getUpgrade(String upgrade) {
+        return Optional.ofNullable(upgrades.get(upgrade));
+    }
+
+    @Override
+    public <T extends IDarkSteelUpgrade> Optional<T> getUpgradeAs(String upgrade, Class<T> as) {
+        IDarkSteelUpgrade up = upgrades.get(upgrade);
+        if (up != null && as.isAssignableFrom(up.getClass())) {
```

Comment:
use class.isInstance, because
```suggestion
        if (as.isInstance(up)) {
```